pandas/core/strings.py

```diff
@@ -999,23 +999,25 @@ def str_get_dummies(arr, sep='|'):
     1  0  0  0
     2  1  0  1
     """
-    arr = arr.fillna('')
-    try:
-        arr = sep + arr + sep
-    except TypeError:
-        arr = sep + arr.astype(str) + sep
+    from pandas.core.reshape.reshape import get_dummies
+    from pandas import Series
+
+    if len(arr) == 0:
+        empty = np.empty(0, dtype='object')
+        return empty, empty
+
+    arr = arr.fillna('').astype('str')
```

Comment:
Why is this astype needed?

Reply:
`astype` is needed as otherwise `Series.str.split` fails when the datatype is not a string (e.g. NaN).

One of the example given in the docstring is:
```
>>> pd.Series(['a|b', np.nan, 'a|c']).str.get_dummies()
a  b  c
0  1  1  0
1  0  0  0
2  1  0  1
```